KAFKA-19541; Allow configuration of kraft RPC max bytes

josefk31 opened the pull request:
Implements KIP-1219 by adding two new configuration properties:
controller.quorum.fetch.snapshot.max.bytes and
controller.quorum.fetch.max.bytes. The
controller.quorum.fetch.snapshot.max.bytes property controls the maximum
bytes for the FETCH_SNAPSHOT request. The
controller.quorum.fetch.max.bytes property controls the maximum bytes
for the FETCH request.

This change also removes the internal.metadata.max.batch.size.in.bytes
internal configuration property. This internal configuration property is
not needed now that there is a publicly available configuration
property.

Reviewers: Alyssa Huang <ahuang@confluent.io>, José Armando García
 Sancio <jsancio@apache.org>

josefk31 committed:
KAFKA-19541 Add new KIP-1219 quorum configurations

Adds two new configurations regulate the amount of data fetched by
Raft clients.

- controller.quorum.fetch.snapshot.max.bytes
- controller.quorum.fetch.max.bytes

Main design choice is that clients control how much they need for
Fetch and FetchSnapshot. The active controller receiving requests will
respect whatever MaxSizeBytes it receives.

ahuang98 commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -1622,7 +1622,7 @@ private FetchResponseData tryCompleteFetchRequest(
             final Records records;
             if (validOffsetAndEpoch.kind() == ValidOffsetAndEpoch.Kind.VALID) {
-                LogFetchInfo info = log.read(fetchOffset, Isolation.UNCOMMITTED);
+                LogFetchInfo info = log.read(fetchOffset, Isolation.UNCOMMITTED, quorumConfig.fetchMaxSizeBytes());
```
> the quorum member initiating the fetch/fetchsnapshot request will decide how much information it desires and the receiving active-controller will respect that wish

hm, from your PR description I would have expected the active controller to read the request to see how many bytes it should return (not for the active controller to read the quorumconfig itself)

josefk31 replied:
Ooops! I should have read this PR again before asking for review :0 But addressed this comment.

ahuang98 commented on raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftLog.java:
```diff
@@ -118,14 +118,22 @@ UnifiedLog log() {
         try {
-            FetchDataInfo fetchInfo = log.read(startOffset, config.internalMaxFetchSizeInBytes(), isolation, true);
+            FetchDataInfo fetchInfo = log.read(startOffset,
+					       defaultMaxTotalRecordsSizeBytes(),
```
this uses the default instead of the maxRecordSizeBytes arg (we should make sure this would have been caught by a test)

josefk31 replied:
Thanks for the comment! Addressed :)

josefk31 commented on raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftLog.java:
```diff
@@ -118,14 +118,22 @@ UnifiedLog log() {
+    }
+
+    @Override
+    public LogFetchInfo read(long startOffset, Isolation readIsolation, int maxRecordSizeBytes) {
```
I think I'll need to rename these variables since it's actually a "Record Batch" which this size applies too!

josefk31 replied:
Done!

ahuang98 commented on raft/src/main/java/org/apache/kafka/raft/RaftLog.java:
```diff
@@ -58,8 +58,26 @@ public interface RaftLog extends AutoCloseable {
+     * Reads which are called from Fetch requests have a configured soft-max at the Quorum level (KIP-1219).
+     * Most reads are, however, intended to be from a disk based raft log to memory.
+     * The default is intended for local reads with only reads intended to be transmitted as a response to fetch using
+     * the non-default implementation.
```
I felt a bit tripped up reading this - "default intended for non-default implementation", "local reads with only reads intended" - wonder if we can phrase this differently

josefk31 replied:
Fair enough. I removed this comment because I felt like there was too much exposition. The most important thing that an implementer should know is that it should be a "soft-max" with respect to batches. Don't think additional information is needed.

ahuang98 commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +93,75 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
 
+    @Test
+    void testFetchRequestObeysConfiguredMaximumBytesToFetch() throws Exception {
+        // Create an explicit test to check that controller.quorum.fetch.max.size.bytes is used to construct fetch
```
did you mean to keep this? maybe we can reword into a description for the test

josefk31 replied:
Done! Let me know what you think :)

ahuang98 commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +93,75 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
+
+        context.pollUntilResponse();
+        FetchResponseData.PartitionData partitionData = context.assertSentFetchPartitionResponse();
+        // Failures for this test will appear in error-logs.
```
what does this mean?

josefk31 replied:
Added comment which is a bit more clear.

ahuang98 commented on raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java:
```diff
@@ -1866,8 +1888,7 @@ FetchRequestData fetchRequest(
         OptionalLong highWatermark,
-        int maxWaitTimeMs
-    ) {
+        int maxWaitTimeMs) {
```
nit, format

josefk31 replied:
Good catch! Done!

ahuang98 commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -1242,7 +1242,7 @@ public void testFetchResponseWithSnapshotId(boolean withKip853Rpc) throws Except
             context.metadataPartition,
             localId,
-            KafkaRaftClient.MAX_FETCH_SIZE_BYTES
+            context.fetchSnapshotMaxSizeBytes
```
similar comment above here that says we validate the maxbytes field?

ahuang98 commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -1321,7 +1323,7 @@ public void testFetchSnapshotResponsePartialData(boolean withKip853Rpc) throws E
             context.metadataPartition,
             localId,
-            KafkaRaftClient.MAX_FETCH_SIZE_BYTES
+            context.fetchSnapshotMaxSizeBytes
```
why is this not `maxFetchSnapshotSizeBytes`?

josefk31 replied:
The kip configuration is `controller.quorum.fetchsnapshot.max.bytes` so it is setup to resemble the configured value (although it is a bit "different" when compared with the old `KafkaRaftClient.MAX_FETCH_SIZE_BYTES`).

ahuang98 replied:
oh, I meant why isn't the value set to the existing `maxFetchSnapshotSizeBytes` variable defined earlier

josefk31 replied:
Woops! Misunderstood comment 🤦 I'll make the change.

ahuang98 commented on raft/src/test/java/org/apache/kafka/raft/internals/KafkaRaftLogTest.java:
```diff
@@ -1083,6 +1086,70 @@ public void testSegmentsLessThanLatestSnapshot() throws IOException {
     }
 
+    @Test
+    public void testReadOfDefaultLogValue() throws IOException {
```
`DefaultLogValue` is a bit generic, what about something like `testReadRespectsDefaultInternalMaxFetchSize`

josefk31 replied:
Great suggestion! Naming things is hard 🤣

ahuang98 commented on raft/src/test/java/org/apache/kafka/raft/internals/KafkaRaftLogTest.java:
```diff
@@ -1083,6 +1086,70 @@ public void testSegmentsLessThanLatestSnapshot() throws IOException {
+        // Exactly 1 batch of records will be read. Since there are 2 batches, with the first batch having 1 record
+        // only 1 record should be returned.
+        assertRecords(info, 1, 1);
+    }
```
what about a case where the first batch has 2 records?

josefk31 replied:
Then 2 records will be returned. It seems like the LogSegment implementation will return at least 1 *batch*. I changed the test slightly to illustrate this.

ahuang98 commented on raft/src/test/java/org/apache/kafka/raft/internals/KafkaRaftLogTest.java:
```diff
@@ -1083,6 +1086,70 @@ public void testSegmentsLessThanLatestSnapshot() throws IOException {
+        append(log, recordsPerBatch, 1);
+        append(log, recordsPerBatch, 1);
+
+        // Default value of 1 is NOT used in this case.
```
NOT used in the case we supply a `maxTotalRecordsSizeBytes`

josefk31 replied:
Good catch! I removed the comment because I think your [suggestion](https://github.com/apache/kafka/pull/21028#discussion_r2688313678) made the code more clear so it is no longer needed.

ahuang98 commented on raft/src/test/java/org/apache/kafka/raft/internals/KafkaRaftLogTest.java:
```diff
@@ -1083,6 +1086,70 @@ public void testSegmentsLessThanLatestSnapshot() throws IOException {
+        assertRecords(info, recordsPerBatch * 2, recordsPerBatch);
+    }
+
+    private static void assertRecords(LogFetchInfo info, int numberExpected, int recordsPerBatch) {
```
java doc!

josefk31 replied:
Hmmm. Since this is a `private static` method and none of the other `private static` utility methods in this class have a java doc I don't think it's really needed.

ahuang98 replied:
I was actually just hoping for a code comment for readers to understand quickly what validation this does (sorry, was being imprecise asking for javadoc)

josefk31 replied:
No problem! Ack'd - I'll add to next iteration :)

ahuang98 commented on raft/src/test/java/org/apache/kafka/raft/internals/KafkaRaftLogTest.java:
```diff
@@ -1083,6 +1086,70 @@ public void testSegmentsLessThanLatestSnapshot() throws IOException {
+        // Default value of 1 is NOT used in this case.
+        LogFetchInfo info = log.read(0,
+                Isolation.UNCOMMITTED,
+                batchSizeBytes * 3);
```
any reason for this value in particular?

josefk31 replied:
Improved the test by parameterizing it and actually worked out an exact meaningful value for `batchSizeBytes`. Hopefully it is more clear.

ahuang98 reviewed (commented):
thanks Jonah! few more questions

josefk31 reviewed (commented):
Thanks for feedback @ahuang98 ! Please let me know if there are any comments which are not sufficiently addressed :)

ahuang98 commented on raft/src/main/java/org/apache/kafka/raft/QuorumConfig.java:
```diff
@@ -117,7 +125,9 @@ public class QuorumConfig {
-            .define(QUORUM_AUTO_JOIN_ENABLE_CONFIG, BOOLEAN, DEFAULT_QUORUM_AUTO_JOIN_ENABLE, LOW, QUORUM_AUTO_JOIN_ENABLE_DOC);
+            .define(QUORUM_AUTO_JOIN_ENABLE_CONFIG, BOOLEAN, DEFAULT_QUORUM_AUTO_JOIN_ENABLE, LOW, QUORUM_AUTO_JOIN_ENABLE_DOC)
+            .define(QUORUM_FETCH_SNAPSHOT_SIZE_MAX_BYTES_CONFIG, INT, DEFAULT_QUORUM_FETCH_SNAPSHOT_SIZE_MAX_BYTES, atLeast(0), LOW, QUORUM_FETCH_SNAPSHOT_SIZE_MAX_BYTES_DOC)
+            .define(QUORUM_FETCH_SIZE_MAX_BYTES_CONFIG, INT, DEFAULT_QUORUM_FETCH_SIZE_MAX_BYTES, atLeast(0), LOW, QUORUM_FETCH_SIZE_MAX_BYTES_DOC);
```
hm, should we allow 0 actually?

josefk31 replied:
Interesting question! Given that we will always return at least 1 batch we _could_ support `0` values. After giving it some thought I am hesitant to do so. I assume that most users will stick with the default or go higher / lower depending on network conditions. 

1. Semantically, it would be confusing to someone reading about this configuration if it were possible to set to `0`. "Oh it's set to 0, that seems like a mistake!"? But admittedly you could say "But setting it to N doesn't really mean setting it to N, it's `max(N, sizeOf(nextBatch))`" so having the possibility of `0` might break peoples assumptions about the value which is maybe good?

2. If we assume that most won't touch this value or have a specific meaningful non-zero intended value it is more likely that a `0` value is a mistake/bug. Consider following example. Many programming languages will automatically `0` integers - for example Golang. If a configuration file for the cluster is generated in `golang` it is possible that there could be a bug which doesn't initiate an integer and sets the configuration to `0`. A counter point to this is that you can also have a bug where the value is `1`, `2`, `3`... or some other small number. But I think the fact that integers are `0` initialized makes it more likely that `0` would be an error rather than intentional. 

3. If someone did just want to, for whatever reason, say "please send at most N bytes or the entire batch" they could get effectively the same effect by setting `1`. 

4. The other configuration will always have a min of `1` byte (admittedly a pathological, yet correct case) so we keep them consistent with each other. 

I think (1) is kinda ambivalent but (2) and (3) combined are a "legit" reason for keeping the allowed value at `1`.

ahuang98 commented on raft/src/test/java/org/apache/kafka/raft/internals/KafkaRaftLogTest.java:
```diff
@@ -1083,6 +1087,74 @@ public void testSegmentsLessThanLatestSnapshot() throws IOException {
+    }
+
+    private static void assertRecordBatches(LogFetchInfo info, int numberExpected, int recordsPerBatch) {
+        // Asserts that we have exactly B * R records. Further there must be B batches of SimpleRecords each with a value of
```
so we're assuming any use case of this validator is with batches with same number of records in them

josefk31 replied:
Yes! That's correct. I don't think we loose much by testing those cases? At least I can't think of any...

josefk31 replied:
Inlined since this function is only used in one other place now.

josefk31 committed:
Renamed test variable to indicate that it is a non-default value

josefk31 committed:
Added an additional test.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/RaftLog.java:
```diff
@@ -58,8 +58,22 @@ public interface RaftLog extends AutoCloseable {
 
     /**
      * Read a set of records within a range of offsets.
+     * maxTotalBatchSizeBytes specifies a "soft" max for total byte size of the record batches to read.
```
Please use Java doc syntax for Java doc comments. Can you add `@param` comments for all of the parameters and `@return` for the returned object.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/RaftLog.java:
```diff
@@ -58,8 +58,22 @@ public interface RaftLog extends AutoCloseable {
+    /**
+     * Configures a soft max for total size of bytes read for {@link #read(long, Isolation)}.
+     */
+    int defaultReadMaxBatchSizeBytes();
```
Why does the client need to know this? Let's try to keep the interface as small as possible.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/RaftLog.java:
```diff
@@ -58,8 +58,22 @@ public interface RaftLog extends AutoCloseable {
+     */
+    default LogFetchInfo read(long startOffsetInclusive, Isolation isolation) {
+        return read(startOffsetInclusive, isolation, defaultReadMaxBatchSizeBytes());
+    }
```
Let's avoid overloaded methods.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftLog.java:
```diff
@@ -118,14 +118,22 @@ UnifiedLog log() {
+            FetchDataInfo fetchInfo = log.read(startOffset,
+                    maxTotalBatchSizeBytes,
+                    isolation,
+                    true);
```
Please use this formatting in the raft module. Notice the comment for the parameter passed using a generic literal like true.

```java
            FetchDataInfo fetchInfo = log.read(
                startOffset,
                maxTotalBatchSizeBytes,
                isolation,
                true // return at least one batch if present
            );
```

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftLog.java:
```diff
@@ -118,14 +118,22 @@ UnifiedLog log() {
-    public LogFetchInfo read(long startOffset, Isolation readIsolation) {
+    public int defaultReadMaxBatchSizeBytes() {
+        return config.internalMaxFetchSizeInBytes();
+    }
```
See my other comment but let's remove this if not needed.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -1517,7 +1517,8 @@ private CompletableFuture<FetchResponseData> handleFetchRequest(
             fetchPartition,
-            currentTimeMs
+            currentTimeMs,
+            request.maxBytes()
```
Let's try to keep the current time as the last parameter. This is consistent with the rest/majority of the methods in this file and module.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -1598,7 +1600,8 @@ private FetchResponseData tryCompleteFetchRequest(
         FetchRequestData.FetchPartition request,
-        long currentTimeMs
+        long currentTimeMs,
+        int maxSizeBytes
```
In the raft module we tend to always make the current time the last parameter of methods that require it.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -3013,7 +3016,7 @@ private FetchSnapshotRequestData buildFetchSnapshotRequest(OffsetAndEpoch snapsh
             quorum.epoch(),
             snapshotId,
-            MAX_FETCH_SIZE_BYTES,
+            quorumConfig.fetchSnapshotSizeMaxBytes(),
```
Why is one configuration call `fetchMaxSizeBytes` while the other is called `fetchSnapshotSizeMaxBytes`? Notice how `Max` and `Size` are swapped.

josefk31 replied:
Good catch. Given that the actual configurations don't mention "size" at all I will just remove any reference to size from either.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/QuorumConfig.java:
```diff
@@ -108,6 +108,14 @@ public class QuorumConfig {
+    public static final int DEFAULT_QUORUM_FETCH_SNAPSHOT_SIZE_MAX_BYTES = 1048576;
+
+    public static final String QUORUM_FETCH_SIZE_MAX_BYTES_CONFIG = QUORUM_PREFIX + "fetch.max.bytes";
+    public static final String QUORUM_FETCH_SIZE_MAX_BYTES_DOC = "Maximum amount of data to retrieve for each Fetch request. Always returns at least one record if a new one is available.";
```
You should mention that FETCH will send at least one batch if there is data to send even if it is greater than QUORUM_FETCH_SIZE_MAX_BYTES_CONFIG.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java:
```diff
@@ -233,7 +233,7 @@ private void checkOffsetIsValid(long offset) {
     private void maybeLoadLog() {
         while (log.endOffset().offset() > nextOffset) {
-            LogFetchInfo info = log.read(nextOffset, Isolation.UNCOMMITTED);
+            LogFetchInfo info = log.read(nextOffset, Isolation.UNCOMMITTED, maxBatchSizeBytes);
```
Okay. Technically this is different from the FETCH case and we don't need to be as strict on how many bytes to read when loading the entire log. This component is attempting to read the entire log and blocks until the entire log is read.

josefk31 replied:
Good catch! Updated this.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/MockLog.java:
```diff
@@ -418,9 +419,29 @@ private void verifyOffsetInRange(long offset) {
+                                expectedSizeBytes, actualSizeBytes));
+            }
+        });
+    }
```
Let's not test this feature this way. You are using this during the protocol test. For those tests you can check that the return fetch response has at most `maxBytesReadFromLog` if the next batch is less than `maxBytesReadFromLog`.

josefk31 replied:
Good suggestion! Wrote another test which does roughly what you are proposing.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java:
```diff
@@ -121,6 +121,8 @@ public final class RaftClientTestContext {
     final int beginQuorumEpochTimeoutMs = fetchTimeoutMs / 2;
     final int retryBackoffMs = Builder.RETRY_BACKOFF_MS;
+    final int fetchSnapshotMaxSizeBytes;
+    final int fetchMaxSizeBytes;
```
It looks like `fetchSnapshotMaxSizeBytes` is not used.

Did you consider making these values parameters to the different methods that need them?

josefk31 replied:
I don't quite understand this comment? We do use it to change the default value in a few tests.

jsancio reviewed (commented):
Thanks for the changes @josefk31 

> I think it's better that clients decide locally what the best conditions are for them

Ideally, we have both. The client communicates how much data they want and the server sends as much data as it can. This implementation is fine for now but ideally kraft should send the min(local limit, remote limit).

> But having two configs with roughly the same function is probably not ideal so I think we should remove the internal.metadata.max.batch.size.in.bytes either in this PR or another one.

Then why not do it in this PR? We should avoid accumulating technical debt.

josefk31 committed:
Responded to CR comments:

- Use number of batches to test KafkaRaftClientFetchTest.java rather
  than custom MockLog error handling
- Use maxBatchSizeBytes as parameter for RaftLog#read in all call sites
- Removed internal configuration for Fetch MaxSizeBytes

josefk31 committed:
Fixed mistake made while resolving merge conflict...

josefk31 committed:
Removed size from new configuration names. Addressed other PR comments

josefk31 commented on raft/src/main/java/org/apache/kafka/raft/QuorumConfig.java:
```diff
@@ -108,6 +108,14 @@ public class QuorumConfig {
+    public static final int DEFAULT_QUORUM_FETCH_SNAPSHOT_MAX_BYTES = 1048576;
+
+    public static final String QUORUM_FETCH_MAX_BYTES_CONFIG = QUORUM_PREFIX + "fetch.max.bytes";
+    public static final String QUORUM_FETCH_MAX_BYTES_DOC = "Maximum amount of data to retrieve for each Fetch request. Always returns at least one batch even if it is greater than QUORUM_FETCH_MAX_BYTES_CONFIG.";
```
I need to change this to `controller.quorum.fetch.max.bytes` or reword it to remove.

jsancio replied:
Is this comment resolved?

josefk31 commented on raft/src/test/java/org/apache/kafka/raft/internals/KafkaRaftLogTest.java:
```diff
@@ -1083,23 +1081,66 @@ public void testSegmentsLessThanLatestSnapshot() throws IOException {
-                MetadataLogConfig.INTERNAL_METADATA_MAX_FETCH_SIZE_IN_BYTES_CONFIG, internalMaxFetchSizeInBytes,
                 MetadataLogConfig.INTERNAL_METADATA_DELETE_DELAY_MILLIS_CONFIG, ServerLogConfigs.LOG_DELETE_DELAY_MS_DEFAULT
-                );
+        );
```
I'll consider undoing this indentation.

josefk31 commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -1023,6 +1023,63 @@ public void testFetchSnapshotRequestAsFollower(boolean withKip853Rpc) throws IOE
     }
 
+    @ParameterizedTest
+    @ValueSource(booleans = { false, true })
```
I'm not sure whether this actually needs to be a parameterized test because `withKip853Rpc` should not have an effect on the snapshot fetching logic. But the other tests did have this parameterization so I stuck with the pattern.

josefk31 commented on raft/src/test/java/org/apache/kafka/raft/MockLog.java:
```diff
@@ -444,7 +445,11 @@ public LogFetchInfo read(long startOffset, Isolation isolation) {
+            if (batch.lastOffset() >= startOffset
+                    && batch.lastOffset() < maxOffset
+                    && !batch.entries.isEmpty()
+                    && (bufferSizeBytes - buffer.remaining()) < maxTotalBatchSizeBytes) {
```
Checks the amount of space we have left in the buffer. Was not able to figure out a better way to do this beyond "max size of buffer - remaining space"

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftLog.java:
```diff
@@ -118,14 +119,19 @@ UnifiedLog log() {
 
         try {
-            FetchDataInfo fetchInfo = log.read(startOffset, config.internalMaxFetchSizeInBytes(), isolation, true);
+            FetchDataInfo fetchInfo = log.read(
+                    startOffset,
+                    maxTotalBatchSizeBytes,
+                    isolation,
+                    true
+            );
```
In Java we indent 4 spaces.

```java
            FetchDataInfo fetchInfo = log.read(
                startOffset,
                maxTotalBatchSizeBytes,
                isolation,
                true
            );
```

This comment applies to a few places.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftLog.java:
```diff
@@ -359,7 +365,12 @@ public Optional<RawSnapshotWriter> createNewSnapshot(OffsetAndEpoch snapshotId)
+        long baseOffset = read(
+                snapshotId.offset(),
+                Isolation.COMMITTED,
+                KafkaRaftClient.MAX_FETCH_SIZE_BYTES
```
Okay but why not ready 0 or 1 byte? Read will guarantee to read at least one batch if one batch exists?

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java:
```diff
@@ -233,7 +234,11 @@ private void checkOffsetIsValid(long offset) {
+            LogFetchInfo info = log.read(
+                    nextOffset,
+                    Isolation.UNCOMMITTED,
+                    KafkaRaftClient.MAX_FETCH_SIZE_BYTES
```
Hmm. I prefer if we just use Integer.MAX_VALUE or something like that. It would allow us to eventually remove `KafkaRaftClient.MAX_FETCH_SIZE_BYTES`. Or in other words can we remove ` KafkaRaftClient.MAX_FETCH_SIZE_BYTES` in this PR?

josefk31 replied:
Could you expand why you have a preference for this? 

~In my mind, it makes sense to leave this unchanged because we're bounding the amount of memory-space that we can allocate in a single read(xxxx) call. After each iteration of the inner loop, the records we have loaded into memory have the potential to be deallocated by GC. But ofc this assumes a more simplistic model for GC and that it can actually free the records but in principal the code loads 8MiB, processes the records and no longer needs them seems reasonable to me. But having no maximum means that the amount of space allocated depends on the size of log on disk. That leads to unpredictable memory usage. We could end up in a situation where there is not enough memory to allocate the entire thing.~

EDIT: Seems like `FileRecord` actually doesn't allocate anything to memory, it seems to just remember the position being read from the file handle...

jsancio commented on raft/src/main/java/org/apache/kafka/raft/QuorumConfig.java:
```diff
@@ -128,6 +138,8 @@ public class QuorumConfig {
     private final int appendLingerMs;
     private final boolean autoJoin;
+    private final int fetchSnapshotSizeMaxBytes;
+    private final int fetchMaxSizeBytes;
```
Please make the names consistent `fetchSnapshotSizeMaxBytes` vs `fetchMaxSizeBytes`. One has `SizeMax` while the other has `MaxSize` based on the configuration name they should be `fetchSnapshotMaxBytes` and `fetchMaxBytes`.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/QuorumConfig.java:
```diff
@@ -177,6 +191,14 @@ public boolean autoJoin() {
         return autoJoin;
     }
 
+    public int fetchSnapshotMaxBytes() {
+        return fetchSnapshotSizeMaxBytes;
+    }
+
+    public int fetchMaxBytes() {
+        return fetchMaxSizeBytes;
+    }
+
```
Make these names consistent with the configuration name used.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/RaftLog.java:
```diff
@@ -58,8 +58,15 @@ public interface RaftLog extends AutoCloseable {
+     * @param startOffsetInclusive Records later than this offset will be returned.
+     * @param isolation Whether to read committed (up to high watermark) or uncommitted data.
+     * @param maxTotalBatchSizeBytes Soft max for number of bytes to retrieve. Will stop returning batches once the
+     *                               size of previously returned batches exceeds maxTotalBatchSizeBytes
```
This is not accurate. Just translate the description from `UnifiedLog#read`:
```       
       * @param maxLength The maximum number of bytes to read
       * @param isolation The fetch isolation, which controls the maximum offset we are allowed to read
       * @param minOneMessage If this is true, the first message will be returned even if it exceeds `maxLength` (if one exists)
```

josefk31 replied:
Reworded the comments a bit.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -1023,6 +1023,63 @@ public void testFetchSnapshotRequestAsFollower(boolean withKip853Rpc) throws IOE
+
+        // Test that we will respond with at least 3 equally sized read of the snapshot.
+        RawSnapshotReader snapshot = context.log.readSnapshot(snapshotId).get();
+        int expectedNumberOfReads = 3;
```
Why 3 reads? I would assume that the invariant is that every fetch snapshot will return at most 6 bytes and that the sum of all of the bytes is equal to the size of the snapshot.

josefk31 replied:
Yeah, took second look and improved the test. I rewrote it to represent a more realistic scenario where it fetches most of the snapshot with the maximum size and then a final call which fetches less than the maximum at the end.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -1023,6 +1023,63 @@ public void testFetchSnapshotRequestAsFollower(boolean withKip853Rpc) throws IOE
+
+        RaftClientTestContext context = new RaftClientTestContext.Builder(localId, voters)
+                .appendToLog(snapshotId.epoch(), List.of("a"))
+                .withFetchSnapshotMaxBytes(6)
```
Make 6 a variable so you can use it in the tests below.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java:
```diff
@@ -487,7 +503,9 @@ public RaftClientTestContext build() throws IOException {
-                listener
+                listener,
+                    fetchSnapshotMaxBytes,
+                    fetchMaxBytes
```
inconsistent indentation.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java:
```diff
@@ -535,6 +555,8 @@ private RaftClientTestContext(
         this.metrics = metrics;
         this.externalKRaftMetrics = externalKRaftMetrics;
         this.listener = listener;
+        this.fetchSnapshotMaxBytes = fetchSnapshotMaxBytes;
```
This variable is not used in this file. Why do you need it? Or are you missing a check for fetch snapshot requests?

josefk31 replied:
It's used to record what value was stored in the `quorumConfig`

1. QuorumConfig where it is used to create `fetchSnapshotRequest`
2. We use it to check fetchSnapshotRequests to see if they are using the configured value where appropriate

The `RaftClientTestContext` seems to have a pattern of storing whatever variables were written to the `quorumConfig` as finalized fields. It does not itself store a reference to the `quorumConfig`. So I think that's why I ended up going in this direction.

jsancio replied:
Right but that is used or done by the builder RaftClientTestContext.Builder and not by RaftClientTestContext itself.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/MockLog.java:
```diff
@@ -428,7 +428,8 @@ public LogFetchInfo read(long startOffset, Isolation isolation) {
 
-        ByteBuffer buffer = ByteBuffer.allocate(512);
+        int bufferSizeBytes = 512;
+        ByteBuffer buffer = ByteBuffer.allocate(bufferSizeBytes);
```
This is not correct now that max total is configurable, no? It may be easiest to use `ByteBufferOutputStream` and check the size after every batch write. Truncating to `max(maxTotalBatchSizeBytes, firstBatchSize)`.

This will simplify the changes to the if statement you made below.

Let's also write test cases for this new functionality.

josefk31 replied:
Hmmm. The original implementation is a bit strange. 
1. It assumes that two batch entries can be written to the 512byte buffer
2. It will only ever return a maximum of two entries (see L#461).

So my implementation idea was "oh, lets keep the behaviour the same but just add the maxTotalBatchSize on top of that".

josefk31 replied:
Did a bit more digging into how the `MockLog` works. 

```
buffer = batch.writeTo(buffer);
```

The proceeding line ends up wrapping buffer in its own `BufferOutputStream` and then dynamically resizing it, returning the resized buffer to caller. 512 is actually the initial buffer size. My initial implementation was incorrect and assumed a fixed size buffer was allocated. To get the correct contract behaviour, we merely need to "bail" if `position() > maxTotalBatchBytes`. I add two tests to document this behaviour.

jsancio reviewed (commented):
Let's update the PR description so that it matches the current implementation.

josefk31 committed:
Addressed PR comments

1. Variable renaming and indentation fixes
2. Rewrote one of the tests to be more sane

josefk31 committed:
Corrected MockLog implementation

- Added tests for MockLog with limits or max size
- Removed KafkaRaftClient.MAX_FETCH_SIZE_BYTES and replaced with Integer.MAX_VALUE

josefk31 committed:
Reverted unneeded change

josefk31 commented on raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java:
```diff
@@ -233,7 +233,11 @@ private void checkOffsetIsValid(long offset) {
 
     private void maybeLoadLog() {
         while (log.endOffset().offset() > nextOffset) {
```
I wonder if we still need this outer loop if we're just going to read everything in a single pass with `log.read(..., Integer.MAX_VALUE)`...

jsancio replied:
The issue is that log.read doesn't say, or guarantee, that it will return everything.

josefk31 commented on raft/src/test/java/org/apache/kafka/raft/MockLog.java:
```diff
@@ -444,7 +444,10 @@ public LogFetchInfo read(long startOffset, Isolation isolation) {
             // batch returned in a fetch response.
-            if (batch.lastOffset() >= startOffset && batch.lastOffset() < maxOffset && !batch.entries.isEmpty()) {
+            if (batch.lastOffset() >= startOffset
+                    && batch.lastOffset() < maxOffset
```
Fixing indentation.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftLog.java:
```diff
@@ -359,7 +364,12 @@ public Optional<RawSnapshotWriter> createNewSnapshot(OffsetAndEpoch snapshotId)
+        long baseOffset = read(
+            snapshotId.offset(),
+            Isolation.COMMITTED,
+            Integer.MAX_VALUE
```
Can we not ask for Integer.MAX_VALUE? If you look at the call it only needs to read the first batch. How about reading 1 byte? Since read will return at least one batch if it exists. This will guarantee that it will read the smallest number of bytes.

josefk31 replied:
Ah! Yes, that is a good idea!

josefk31 committed:
Fixed indentation to 8->4 spaces

jsancio commented on raft/src/test/java/org/apache/kafka/raft/internals/KafkaRaftLogTest.java:
```diff
@@ -1083,23 +1081,66 @@ public void testSegmentsLessThanLatestSnapshot() throws IOException {
+                    record.timestamp(),
+                    record.key(),
+                    record.value(),
+                    record.headers()
```
Since this is testing that the max bytes parameter is satisfied do we need to test the behavior of simple record? Why not just assert that `assertEquals(expectedValue, record.value())`

josefk31 replied:
Was unaware of `ByteBuffer.wrap(xxxx)` - fixed.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/internals/KafkaRaftLogTest.java:
```diff
@@ -1083,23 +1081,66 @@ public void testSegmentsLessThanLatestSnapshot() throws IOException {
+        append(log, recordsPerBatch, 1);
+        append(log, recordsPerBatch, 1);
+
+        LogFetchInfo info = log.read(
```
Why not check the size in bytes of the records returned?

josefk31 replied:
Done!

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +95,77 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
     }
 
+    @Test
+    void testSentFetchUsesQuorumMaxBytesConfiguration() throws Exception {
```
This is a strange test since the check is implicit in assertSentFetchReqest. It is trusting that the test context is setting and checking the request field correctly, no?

josefk31 replied:
Good point. Made tests more explicit.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +95,77 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
+                )
+                .withUnknownLeader(epoch)
+                .withFetchMaxBytes(localMaxSizeBytes)
+                .build();
```
Please fix indentation. We indent by 4 spaces in Java and the raft module.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +95,77 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
+        // There are two batches with 3 records each. The first batch is always at least larger than 1 byte.
+        // If remoteMaxSizeBytes = 1 then the MockLog will return exactly 1 batch.
+        // If localMaxSizeBytes is used then MockLog will return two batches
+        var remoteMaxSizeBytes = 1;
+        var localMaxSizeBytes = 1024;
+
+        RaftClientTestContext context = new RaftClientTestContext.Builder(
+                localKey.id(),
+                localKey.directoryId().get()
+        )
+                .appendToLog(epoch, List.of("a", "a", "a"))
+                .appendToLog(epoch, List.of("b", "b", "b"))
+                .withStartingVoters(
+                        VoterSetTest.voterSet(Stream.of(localKey, remoteKey)), KRaftVersion.KRAFT_VERSION_1
+                )
+                .withUnknownLeader(epoch)
+                .withFetchMaxBytes(localMaxSizeBytes)
+                .build();
+
+        context.unattachedToLeader();
+        epoch = context.currentEpoch();
+
+        // Send a fetch request with max bytes that are different from the configured value.
+        FetchRequestData request = context.fetchRequest(epoch, remoteKey, 1L, epoch, 500);
+        request.setMaxBytes(remoteMaxSizeBytes);
+        context.deliverRequest(request);
+
+        context.pollUntilResponse();
+        FetchResponseData.PartitionData partitionData = context.assertSentFetchPartitionResponse();
+        assertEquals(Errors.NONE.code(), partitionData.errorCode());
+        MemoryRecords records = (MemoryRecords) FetchResponse.recordsOrFail(partitionData);
+        var iterator = records.batchIterator();
+        var firstBatch = iterator.next();
+        assertEquals(0, firstBatch.baseOffset());
+        assertEquals(3, firstBatch.nextOffset());
+        assertFalse(iterator.hasNext(), "maxSize=1 implies a single batch is retrieved");
+    }
+
+
+
```
Extra spaces.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +95,77 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
+    }
+
+    @Test
+    public void testMaxBytesRequestedFromLogsUsesValueFromFetch() throws Exception {
```
What about the other case? Where Fetch returns up to remoteMaxSizeBytes if there are that many bytes and the first batch is less that remoteMaxSizeBytes?

jsancio commented on raft/src/test/java/org/apache/kafka/raft/internals/KafkaRaftLogTest.java:
```diff
@@ -1083,23 +1081,66 @@ public void testSegmentsLessThanLatestSnapshot() throws IOException {
+    public void testReadRespectsMaxSizeInBytes(int expectedBatches) throws IOException {
+        // 5 records are written in batches of 101 bytes each (at time of writing).
+        int magicMaxBatchSizeBytes = 101;
+        MetadataLogConfig config = createMetadataLogConfig(
+                10240,
+                10 * 1000,
+                10240,
+                60 * 1000,
+                magicMaxBatchSizeBytes
+        );
```
Please fix indentation.

josefk31 replied:
8 spaces for continuations is consistent with the rest of the file. I'd prefer to just create a "whitespace" specific PR for this.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -1023,6 +1023,86 @@ public void testFetchSnapshotRequestAsFollower(boolean withKip853Rpc) throws IOE
+                .appendToLog(snapshotId.epoch(), List.of("a"))
+                .withFetchSnapshotMaxBytes(fetchSnapshotMaxBytes)
+                .withKip853Rpc(true)
+                .build();
```
Please fix indentation.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -1023,6 +1023,86 @@ public void testFetchSnapshotRequestAsFollower(boolean withKip853Rpc) throws IOE
+        // When a snapshot is written there is usually overhead "control" bytes.
+        // At time of writing the size of snapshot is 237 bytes.
+        // In the test scenario, we will fetch the max snapshot size 2 times and then fetch the remaining bits.
+        int fetchSnapshotMaxBytes = 118;
```
This all seem very magical. Why not discover these values by reading the size of the generated snapshot?

josefk31 replied:
Fixed! ~But it's kinda hard to actually set this value dynamically so I'm not sure if we're getting to a better solution...~ 

EDIT: Found an easier way.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -1023,6 +1023,86 @@ public void testFetchSnapshotRequestAsFollower(boolean withKip853Rpc) throws IOE
+        int totalBytesRead = 0;
+        int position = 0;
+        for (int i = 0; i < expectedNumberOfReads; i++) {
+            context.deliverRequest(
+                    fetchSnapshotRequest(
+                            context.metadataPartition,
+                            epoch,
+                            snapshotId,
+                            fetchSnapshotMaxBytes,
+                            position
+                    )
+            );
```
Please fix indentation.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -1023,6 +1023,86 @@ public void testFetchSnapshotRequestAsFollower(boolean withKip853Rpc) throws IOE
+        // Test that we will respond with at least 2 equally sized read of the snapshot.
+        RawSnapshotReader snapshot = context.log.readSnapshot(snapshotId).get();
+        int snapshotSizeBytes = Math.toIntExact(snapshot.sizeInBytes());
+        int expectedNumberOfReads = 2;
```
Why not compute this based on the snapshot size and the size of the requests?

josefk31 replied:
Addressed while addressing related comment.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -1023,6 +1023,86 @@ public void testFetchSnapshotRequestAsFollower(boolean withKip853Rpc) throws IOE
+        // Fetch the remaining snapshot bytes.
+        assertTrue(
+            totalBytesRead < snapshotSizeBytes,
+            "Total bytes read should be less than snapshot size");
```
How about this indentation?

```java
        assertTrue(
            totalBytesRead < snapshotSizeBytes,
            "Total bytes read should be less than snapshot size"
        );
```

I suggest printing the total bytes read and snapshot size bytes in the assert true message.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -1023,6 +1023,86 @@ public void testFetchSnapshotRequestAsFollower(boolean withKip853Rpc) throws IOE
+        assertTrue(
+            totalBytesRead < snapshotSizeBytes,
+            "Total bytes read should be less than snapshot size");
+        context.deliverRequest(
+                fetchSnapshotRequest(
+                        context.metadataPartition,
+                        epoch,
+                        snapshotId,
+                        fetchSnapshotMaxBytes,
+                        position
+                )
+        );
```
Please fix indentation.

josefk31 committed:
Addressed PR feedback

1. Removed fetchSnapshotMaxBytes from KafkaRaftTestContext
2. Indentation fixes
3. For testFetchSnapshotRequestWithPartialData, figure out how many requests to send based on snapshot size

josefk31 committed:
Responded to PR feedback

1. Improved tests
2. Added additional test for fetch
3. Made check for batchSize more explicit

josefk31 committed:
Added additional tests to address PR feedback

jsancio commented on raft/src/test/java/org/apache/kafka/raft/MockLog.java:
```diff
@@ -444,6 +444,7 @@ public LogFetchInfo read(long startOffset, Isolation isolation, int maxTotalBatc
             // complete batches, so batches which end at an offset larger than the max offset are
             // filtered, which is effectively the same as having the consumer drop an incomplete
             // batch returned in a fetch response.
+            // Exits loop if position of buffer goes beyond maxTotalBatchBytes.
```
Ideally a good comment explains the "why" behind the implementation, not the "what" or "how". The reader can understand the what or how by reading the code. E.g. why does the code need to exit the loop when the buffer position is beyond max total batch bytes?

jsancio commented on raft/src/main/java/org/apache/kafka/raft/RaftLog.java:
```diff
@@ -57,9 +57,16 @@ public interface RaftLog extends AutoCloseable {
+     * @return Records and start offset information wrapped in a LogFetchInfo
      */
-    LogFetchInfo read(long startOffsetInclusive, Isolation isolation);
+    LogFetchInfo read(long startOffsetInclusive, Isolation isolation, int maxTotalBatchBytes);
```
Did you take a look at the name and description used by UnifiedLog? How about:

```java
    /**
     * @param maxLength the maximum number of bytes to read if there are more than one records batch
     */
    LogFetchInfo read(long startOffsetInclusive, Isolation isolation, int maxLength);
```

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftLog.java:
```diff
@@ -359,7 +364,12 @@ public Optional<RawSnapshotWriter> createNewSnapshot(OffsetAndEpoch snapshotId)
+        long baseOffset = read(
+            snapshotId.offset(),
+            Isolation.COMMITTED,
+            1
```
Try to add a comment explaining the paramster when using literals.
```java
            1  // maxLength
```

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +96,191 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
+            local.directoryId().get()
+        )
+            .withStartingVoters(
+                VoterSetTest.voterSet(Stream.of(local, electedLeader)), KRaftVersion.KRAFT_VERSION_1
```
How about:
```java
                VoterSetTest.voterSet(Stream.of(local, electedLeader)),
                KRaftVersion.KRAFT_VERSION_1
```

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +96,191 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
+            .appendToLog(epoch, List.of("a", "a", "a"))
+            .appendToLog(epoch, List.of("b", "b", "b"))
+            .withStartingVoters(
+                VoterSetTest.voterSet(Stream.of(localKey, remoteKey)), KRaftVersion.KRAFT_VERSION_1
```
How about:

```java
                VoterSetTest.voterSet(Stream.of(localKey, remoteKey)),
                KRaftVersion.KRAFT_VERSION_1
```

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +96,191 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
+            .appendToLog(epoch, List.of("a", "a", "a"))
+            .appendToLog(epoch, List.of("b", "b", "b"))
+            .withStartingVoters(
+                VoterSetTest.voterSet(Stream.of(localKey, remoteKey)), KRaftVersion.KRAFT_VERSION_1
```
Same comment here. Please add a newline between the parameters. Long parameter list in one line are hard to read.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +96,191 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
+            .appendToLog(epoch, List.of("b", "b", "b"))
+            .appendToLog(epoch, List.of("c", "c", "c"))
+            .withStartingVoters(
+                VoterSetTest.voterSet(Stream.of(localKey, remoteKey)), KRaftVersion.KRAFT_VERSION_1
```
Same here.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +96,191 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
+    }
+
+    @Test
+    public void testFetchMaxBytesFromRemoteFetchUsed() throws Exception {
```
Aren't all of the tests testing this? "remote fetch used"? To me this test is testing `testFetchMaxBytesReturnMaxBytesForTwoOrMoreBatches`. I tend to write a short comment to explain the test if the test method name is too cryptic. E.g.

```java
    public void testFetchMaxBytesFromRemoteFetchUsed() throws Exception {
        // test case: FETCH returns at most maxBytes if the first batch is less than maxBytes.
        ...
```

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +96,191 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
+        MemoryRecords records = (MemoryRecords) FetchResponse.recordsOrFail(partitionData);
+        // Invariant is that we will always return records in batches and will include batches which "go over"
+        // the controller.quorum.fetch.max.bytes configuration.
+        assertTrue(
+            records.sizeInBytes() < remoteMaxSizeBytes * 2,
+            String.format(
+                "Expected records size (%d) < remoteMaxSizeBytes*2 (%d)",
+                records.sizeInBytes(),
+                remoteMaxSizeBytes * 2
+            )
+        );
```
This invariant is not obvious to me. Ideally, you would want this invariant:
```java
(firstBatch.sizeInBytes() == records.sizeInBytes() && records.sizeInBytes() >= remoteMaxSizeBytes) || records.sizeInBytes() <= remoteMaxSizeBytes
```
or this more general invariant
```java
firstBatch.sizeInBytes() == records.sizeInBytes() || records.sizeInBytes() <= remoteMaxSizeBytes
```

josefk31 replied:
Hmmm. I think the implementation doesn't actually do this invariant. It will always go a bit over the max because it's returning batches unless the batch is exactly the same size as the `remoteMaxSizeBytes`.

jsancio replied:
I see. That's for the mock implementation, right? If so, we need to fix that.

josefk31 replied:
Good catch - changed how `MockLog` is implemented.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java:
```diff
@@ -1023,6 +1023,87 @@ public void testFetchSnapshotRequestAsFollower(boolean withKip853Rpc) throws IOE
+        // Test that we will respond with at least 2 equally sized read of the snapshot.
+        RawSnapshotReader snapshot = context.log.readSnapshot(snapshotId).get();
+        int snapshotSizeBytes = Math.toIntExact(snapshot.sizeInBytes());
+        int expectedNumberOfReads = 2;
```
Okay but reading the code it looks like the expected number of FETCH_SNAPSHOT is `expectedNumberOfReads + 1`.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/MockLog.java:
```diff
@@ -513,7 +517,11 @@ public Optional<RawSnapshotWriter> createNewSnapshot(OffsetAndEpoch snapshotId)
+        long baseOffset = read(
+            snapshotId.offset(),
+            Isolation.COMMITTED,
+            KafkaRaftClient.MAX_FETCH_SIZE_BYTES
```
This only needs to read one batch, right? Why not make it size 1 like we do somewhere else?

jsancio commented on raft/src/test/java/org/apache/kafka/raft/MockLogTest.java:
```diff
@@ -991,6 +1041,48 @@ public void testValidateValidEpochAndOffset() {
+        // magicMaxTotalBytes are smaller than 10 simple records in a batch.
+        // Meaning we will read only the first batch and not the second.
+        int magicMaxTotalBytes = 100;
+        Records records = log.read(
+                0,
+                Isolation.UNCOMMITTED,
+                magicMaxTotalBytes
+        ).records;
```
Extra spaces for the indentation.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/MockLogTest.java:
```diff
@@ -991,6 +1041,48 @@ public void testValidateValidEpochAndOffset() {
+            lastReadOffset = record.offset();
+            assertEquals(expectedOffset, lastReadOffset);
+            expectedOffset += 1;
+        }
```
What is this testing? To me this is testing that offset are monotonically increasing, no? And that the last offset is 9, right?

josefk31 replied:
We expect only a monotonically increasing sequence of records `[0...9]` to be returned.

josefk31 replied:
This tests that the method obeys the contract and will return at least one batch. I've changed the test name and added some others to reflect this.

josefk31 replied:
Changed this to be a "count" which is simpler to understand.

jsancio reviewed (commented):
Thanks for the detail tests.

josefk31 committed:
Feedback response:

1. New test for MockLog
2. Indentation fixes
3. Comment fixes

josefk31 committed:
Altered a few more test invariants

josefk31 committed:
Updated comments

jsancio commented on raft/src/test/java/org/apache/kafka/raft/MockLog.java:
You need to truncate the buffer to max total batch bytes, if the first batch is less than max total batch bytes.

josefk31 replied:
I used slice - which I believe is the correct behaviour.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/MockLogTest.java:
```diff
@@ -991,6 +1041,89 @@ public void testValidateValidEpochAndOffset() {
+        // MockLog#read returns data in batches and will return an additional batch if one of them
+        // exceeds maxTotalBytes.
+        // The sum of the size of both batches is expected to be greater than
+        // magicMaxTotalBytes but less than 2 x magicMaxTotalBytes
```
I see? Why is that? See my comment but we can truncate the buffer to max total bytes if it read more than one batch. This will match the src/main implementation.

josefk31 committed:
Changed mocklog invariant and updated tests accordingly

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -261,33 +269,32 @@ public void testFetchMaxBytesOneOrMoreBatches() throws Exception {
+        // If we return 2 or more batches, invariant is to always return less than remoteMaxSizeBytes
         assertTrue(
-            records.sizeInBytes() < remoteMaxSizeBytes * 2,
+            records.sizeInBytes() < remoteMaxSizeBytes,
```
This should be `<=`, no?

josefk31 replied:
I added an additional test which tests the `=` invariant. In this test, I think `<` is expected.

jsancio replied:
Right but the comment above says "invariant is to return <= remoteBatchSizeBytes."

jsancio commented on raft/src/test/java/org/apache/kafka/raft/MockLog.java:
```diff
@@ -468,6 +468,9 @@ public LogFetchInfo read(long startOffset, Isolation isolation, int maxTotalBatc
 
         buffer.flip();
         Records records = MemoryRecords.readableRecords(buffer);
+        if (batchCount > 1) {
+            records = records.slice(0, maxTotalBatchBytes);
+        }
```
Doesn't `maxTotalBatchBytes` need to be less than the capacity? How about:
```java
if (batchCount > 1 && maxTotalBatchBytes < records.limit()) {
    records.limit(maxTotalBatchBytes);
}
```

josefk31 replied:
Hmm, there is no `records.limit()` field - it seems like `slice` is the only way exposed to truncate.

josefk31 replied:
Did more digging, I think we avoid problems here because of:

https://github.com/apache/kafka/blob/0166a0342d82b3b9b89397b15304e974507c7eb8/clients/src/main/java/org/apache/kafka/common/record/internal/MemoryRecords.java#L313

The `MemoryRecords` will typically truncate to the buffer limit.

josefk31 committed:
Added more clear test for batchSize invariants

josefk31 committed:
Added additional test for slice truncation behaviour

jsancio commented on raft/src/test/java/org/apache/kafka/raft/MockLogTest.java:
```diff
@@ -1105,7 +1126,7 @@ public void testMockLogReadExtremelyLargeInitialBatch() {
             iterator.next();
         }
-        assertEquals(numberOfRecords, recordCount);
+        assertEquals(numberOfRecords, recordCount * 2);
```
Why? This reading 1 byte so I would assume that it would read only one batch which has numberOfRecords record.

josefk31 replied:
Made a mistake and did not run the test before pushing. The test actually fails with this value. Reverted this change.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/MockLogTest.java:
```diff
@@ -1090,9 +1090,30 @@ public void testMockLogLimitsReturnsAtLeastOne() {
+            recordCount++;
+            iterator.next();
+        }
+        assertEquals(numberOfRecords * 2, recordCount);
```
Please explain why. I am having a hard time understanding what it is being tested. It is important to distinguish between the semantic of the API and issues with the implementation.

Since the read call is for MAX_VALUE, it should read all of the batches and records which is "number of records times 2" since there are two batches with number of records each, right?

josefk31 replied:
Yeah, that's correct. Let me rephrase the variables names a little.

josefk31 committed:
Fixed broken test and clarified variable names

jsancio commented on raft/src/test/java/org/apache/kafka/raft/internals/KafkaRaftLogTest.java:
```diff
@@ -1083,23 +1081,58 @@ public void testSegmentsLessThanLatestSnapshot() throws IOException {
+
+    @ParameterizedTest
+    @ValueSource(ints = {1, 2, 3})
+    public void testReadRespectsMaxSizeInBytes(int expectedBatches) throws IOException {
```
I should have mentioned this earlier but I just remembered that KafkaRaftLog and MockLog implement the same interface. They should have the same semantic and tests. Ideally we would have the same tests running against both implementation. Until we get there, can you make sure that we have the same new tests in KafkaRaftLogTest and MockLogTest.

For example, you added one test to KafkaRaftLogTest and three tests to MockLogTest.

josefk31 replied:
Good idea! There was one set of tests about large batch sizes in `MockLog` that doesn't make sense to move. But Moved the others which test a "semantic" meaning.

josefk31 committed:
Made tests similar in MockLog and KafkaRaftLog

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +96,288 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
+        var remoteMaxSizeBytes = 1;
+        var localMaxSizeBytes = 1024;
+
+        RaftClientTestContext context = new RaftClientTestContext.Builder(
+                localKey.id(),
+                localKey.directoryId().get()
+        )
```
Extra 4 spaces.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +96,288 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
+    }
+
+    @Test
+    public void testFetchMaxBytesOneOrMoreBatches() throws Exception {
```
This is testing two or more batches, no? Yet the test name is "..OneOrMoreBatches"

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +96,288 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
+        FetchRequestData exactSizeBytesRequest =
+            context.fetchRequest(epoch, remoteKey, 1L, epoch, 500);
+        exactSizeBytesRequest.setMaxBytes(allRecords.sizeInBytes());
+        context.deliverRequest(allRecordsRequest);
```
This is sending allRecordsRequest and not exactSizeBytesRequest. Given that the test pass what is this actually testing?

josefk31 replied:
Good catch! Fixed!

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +96,288 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
+        assertFalse(allIterator.hasNext());
+
+        // Here we sent a fetch with sizeInBytes-1, this will end up returning a single batch
+        // The other batch is omitted.
```
This is not completely accurate. FETCH response should have sent all of the requested bytes. What you are observing is that, as expected, MemoryRecords cannot parse those bytes because they are not enough bytes.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientFetchTest.java:
```diff
@@ -91,6 +96,288 @@ private static void testFetchResponseWithInvalidRecord(MemoryRecords records, in
+        context.pollUntilResponse();
+        MemoryRecords oneBatchRecords =
+            (MemoryRecords) FetchResponse.recordsOrFail(context.assertSentFetchPartitionResponse());
+        assertTrue(oneBatchRecords.sizeInBytes() < exactSizeBytesRecords.sizeInBytes());
```
This is true because `oneBatchRecords.sizeInBytes == exactSizeBytesRecords.sizeInBytes() - 1` and `exactSizeBytesRecords.sizeInBytes() - 1 < exactSizeBytesRecords.sizeInBytes()` by definition. This is not really an invariant.

As I mentioned earlier to me the invariant is:
```java
firstBatch.sizeInBytes() == records.sizeInBytes() || records.sizeInBytes() <= remoteMaxSizeBytes
```

josefk31 replied:
Added some additional assertions for invariants in addition to those which already exist in test.

josefk31 committed:
Fixed comments on tests

josefk31 committed:
Addressed comment

jsancio commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
Nice, this symbol is not used anymore. Can we remove this symbol?

josefk31 replied:
Removed :)

josefk31 committed:
Removed unused MAX_FETCH_SIZE_BYTES

jsancio reviewed (approved):
LGTM
